{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "575d87d1",
   "metadata": {},
   "source": [
    "https://www.yeastgenome.org/\n",
    "\n",
    "https://www.cse.msu.edu/~cse835/Papers/Graph_connectivity_revised.pdf => connexite\n",
    "\n",
    "Maslov et al., 2013, Segundo et al., 2011, Tomita and Kameda, 2007 => Ensemble stable\n",
    "\n",
    "\n",
    "isthme, trouver les cycles et circonférence, sommet-connexite, centralité, nb chromatique, contraction plutôt que suppression, partition, dureté, force, stable, clique, noyau, dominants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f20180f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6f88050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parcours_profondeur(graphe: dict, sommet: int) -> list:\n",
    "    \"\"\"\n",
    "        Algorithme de parcours en profondeur d'un graphe.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "            - sommet (int): correspond à la racine de l'aborescence de parcours en profondeur.\n",
    "        \n",
    "        Sorties:\n",
    "            - pere (list): correspon à l'arborescence avec pere[x] est le père du sommet x dans le parcous en profondeur.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n)\n",
    "    \"\"\"\n",
    "    taille = max(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    pere = [None for _ in range(taille)]\n",
    "    dernier_pred = [None for _ in range(taille)]\n",
    "    \n",
    "    nb_vu = 1\n",
    "    \n",
    "    pile = [sommet]\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        \n",
    "        if nb_vu not in graphe.keys():\n",
    "            \n",
    "            nb_vu += 1\n",
    "        \n",
    "        elif etat[nb_vu - 1] is not None:\n",
    "            \n",
    "            nb_vu += 1\n",
    "            \n",
    "        elif pile == []:\n",
    "            \n",
    "            pile.append(nb_vu)\n",
    "            \n",
    "        while pile != []:\n",
    "            \n",
    "            suivant = pile.pop()\n",
    "            \n",
    "            if etat[suivant - 1] is None:\n",
    "                \n",
    "                etat[suivant - 1] = 0\n",
    "                pere[suivant - 1] = dernier_pred[suivant - 1]\n",
    "                \n",
    "                for successeur in graphe[suivant]:\n",
    "                    \n",
    "                    pile.append(successeur)\n",
    "                    dernier_pred[successeur - 1] = suivant\n",
    "    return pere\n",
    "\n",
    "\n",
    "def excentricite_sommet(graphe: dict, sommet: int) -> int:\n",
    "    \"\"\"\n",
    "        Algorithme de calcul d'excentricité d'un sommet dans un graphe. Aussi la hauteur de l'aborescence d'un parcours\n",
    "        en largeur du graphe à partir du sommet.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "            - sommet (int): correspond à la racine de l'aborescence de parcours en profondeur.\n",
    "        \n",
    "        Sorties:\n",
    "            - max(hauteur) (int): correspond à l'excentricité du sommet.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n)\n",
    "    \"\"\"\n",
    "    taille = max(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    hauteur = [0 for _ in range(taille)]\n",
    "    \n",
    "    file = [sommet]\n",
    "    etat[sommet - 1] = 0\n",
    "    \n",
    "    nb_vu = 1\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        \n",
    "        if nb_vu not in graphe.keys():\n",
    "            \n",
    "            nb_vu += 1\n",
    "        \n",
    "        elif etat[nb_vu - 1] is not None:\n",
    "            \n",
    "            nb_vu += 1\n",
    "            \n",
    "        elif file == []:\n",
    "            \n",
    "            file.append(nb_vu)\n",
    "            etat[nb_vu - 1] = 0\n",
    "            \n",
    "        while file != []:\n",
    "            \n",
    "            suivant = file[0]\n",
    "            file = file[1:] if len(file) > 1 else []\n",
    "            \n",
    "            for successeur in graphe[suivant]:\n",
    "                \n",
    "                if etat[successeur - 1] is None:\n",
    "                    \n",
    "                    file.append(successeur)\n",
    "                    etat[successeur - 1] = 0\n",
    "                    hauteur[successeur - 1] = hauteur[suivant - 1] + 1\n",
    "                    \n",
    "    return max(hauteur)\n",
    "\n",
    "\n",
    "def composantes_connexes(graphe: dict) -> list:\n",
    "    \"\"\"\n",
    "        Algorithme de calcul des composantes connexes d'un graphe par parcours en profondeur, à chaque fois qu'un sommet n'a\n",
    "        pas était visité alors que la pile est vide, compose connexe + 1.\n",
    "        L'algorithme renvoie de plus les composantes connexes sous forme de liste avec un dictionnaire pour représenter chaque\n",
    "        composante connexe.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties: \n",
    "            - composante (list): Les composantes connexes du graphe.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n).\n",
    "    \"\"\"\n",
    "    taille = max(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    pere = [None for _ in range(taille)]\n",
    "    dernier_pred = [None for _ in range(taille)]\n",
    "    composante = []\n",
    "    \n",
    "    nb_vu = 1\n",
    "    sous_graphe_connexe = {1: set()}\n",
    "    \n",
    "    pile = [1]\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        \n",
    "        if nb_vu not in graphe.keys():\n",
    "            \n",
    "            nb_vu += 1\n",
    "        \n",
    "        elif etat[nb_vu - 1] is not None:\n",
    "            \n",
    "            nb_vu += 1\n",
    "            \n",
    "        elif pile == []:\n",
    "            \n",
    "            composante.append(sous_graphe_connexe)\n",
    "            pile.append(nb_vu)\n",
    "            sous_graphe_connexe = {nb_vu: set()}\n",
    "            \n",
    "        while pile != []:\n",
    "            \n",
    "            suivant = pile.pop()\n",
    "            \n",
    "            if etat[suivant - 1] is None:\n",
    "                \n",
    "                sous_graphe_connexe[suivant] = set()\n",
    "                etat[suivant - 1] = 0\n",
    "                pere[suivant - 1] = dernier_pred[suivant - 1]\n",
    "                \n",
    "                for successeur in graphe[suivant]:\n",
    "                    \n",
    "                    sous_graphe_connexe[successeur] = set() if etat[successeur - 1] is None else sous_graphe_connexe[successeur]\n",
    "                    sous_graphe_connexe[suivant].add(successeur)\n",
    "                    sous_graphe_connexe[successeur].add(suivant)\n",
    "                    pile.append(successeur)\n",
    "                    dernier_pred[successeur - 1] = suivant\n",
    "                    \n",
    "    composante.append(sous_graphe_connexe)\n",
    "    \n",
    "    return composante\n",
    "\n",
    "\n",
    "def sommets_isoles(graphe: dict) -> set:\n",
    "    \"\"\"\n",
    "        Algorithme qui renvoie les sommets isolées d'un graphe, ou les sommets de degré = 0.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "            \n",
    "        Sorties:\n",
    "            - isoles (set): ensemble de sommets isolées i.e de degré nul.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n)\n",
    "    \"\"\"\n",
    "    isoles = set()\n",
    "    \n",
    "    for v1, voisins in graphe.items():\n",
    "        \n",
    "        if len(voisins) == 0:\n",
    "            \n",
    "            isoles = isoles.union({v1})\n",
    "    \n",
    "    return isoles\n",
    "\n",
    "\n",
    "def cycle(graphe: dict) -> bool:\n",
    "    \"\"\"\n",
    "        Algorithme qui réalise un parcours en profondeur, si rencontre deux fois un sommet alors il renvoie vrai \n",
    "        si le graphe possède un cycle, faux sinon.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - cyclique (bool): Vrai si le graphe possède un cycle, faux sinon.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n)\n",
    "    \"\"\"\n",
    "    cyclique = False\n",
    "    \n",
    "    taille = len(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    dernier_pred = [None for _ in range(taille)]\n",
    "    \n",
    "    nb_vu = 1\n",
    "    \n",
    "    pile = [list(graphe.keys())[0]]\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        \n",
    "        if nb_vu not in graphe.keys():\n",
    "            \n",
    "            nb_vu += 1\n",
    "        \n",
    "        elif etat[nb_vu - 1] is not None:\n",
    "            \n",
    "            nb_vu += 1\n",
    "            \n",
    "        elif pile == []:\n",
    "            \n",
    "            pile.append(nb_vu)\n",
    "            cc += 1\n",
    "            \n",
    "        while pile != []:\n",
    "            \n",
    "            suivant = pile.pop()\n",
    "            \n",
    "            if etat[suivant - 1] is None:\n",
    "                \n",
    "                etat[suivant - 1] = 0\n",
    "                \n",
    "                for successeur in graphe[suivant]:\n",
    "                    \n",
    "                    if (etat[successeur - 1] is not None) and (dernier_pred[suivant - 1] != successeur):\n",
    "                        \n",
    "                        cyclique = True\n",
    "                        break\n",
    "                        \n",
    "                    pile.append(successeur)\n",
    "                    dernier_pred[successeur - 1] = suivant\n",
    "    \n",
    "                if cyclique: break\n",
    "            if cyclique: break\n",
    "        if cyclique: break\n",
    "    \n",
    "    return cyclique\n",
    "\n",
    "\n",
    "def points_articulations(graphe: dict) -> set:\n",
    "    \"\"\"\n",
    "        Algorithme qui renvoie l'ensemble des points d'articulations d'un graphe, par un parcours en profondeur si la racine\n",
    "        possède N fils avec N > 1 alors point d'articulation.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - points (set): L'ensemble des points d'articulations d'un graphe.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n^2)\n",
    "    \"\"\"\n",
    "    points = set()\n",
    "    \n",
    "    for s in graphe:\n",
    "        points = points.union({s}) if parcours_profondeur(graphe, s).count(s) > 1 else points\n",
    "    \n",
    "    return points\n",
    "\n",
    "\n",
    "articulation = lambda graphe, s: True if parcours_profondeur(graphe, s).count(s) > 1 else False\n",
    "\n",
    "\n",
    "def supprimer_sommet(graphe: dict, sommet: int) -> dict:\n",
    "    \"\"\"\n",
    "        Algorithme permettant la création d'un sous-graphe induit avec G' = (V', E') et V - {sommet}.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "            - sommet (int): le sommet a supprimé.\n",
    "        \n",
    "        Sorties:\n",
    "            - nouveau_graphe (dict): correspond au sous-graphe induit.\n",
    "        \n",
    "        Complexité:\n",
    "            - O(|V|+|E|)\n",
    "    \"\"\"\n",
    "    nouveau_graphe = graphe.copy()\n",
    "    \n",
    "    for s in nouveau_graphe.keys():\n",
    "        nouveau_graphe[s] = nouveau_graphe[s].difference({sommet})\n",
    "    del nouveau_graphe[sommet]\n",
    "    \n",
    "    return nouveau_graphe\n",
    "\n",
    "def graphe_complementaire(graphe: dict) -> dict:\n",
    "    \"\"\"\n",
    "        Algorithme qui permet de calculer le graphe complémentaire de G.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - complementaire (dict): graphe complémentaire du graphe d'entrée\n",
    "        \n",
    "        Complexité:\n",
    "            - O(?)\n",
    "    \"\"\"\n",
    "    complementaire = {sommet: set(graphe.keys()) for sommet in graphe.keys()}\n",
    "    \n",
    "    for s, voisins in graphe.items():\n",
    "        \n",
    "        complementaire[s] = complementaire[s].difference(graphe[s].union({s}))\n",
    "\n",
    "    return complementaire\n",
    "\n",
    "    \n",
    "def centres(graphe: dict) -> tuple:\n",
    "    \"\"\"\n",
    "        Algorithme calculant l'ensemble des centres d'un graphe par le calcul de l'excentricité minimale.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - (tuple): Renvoie (x, y) avec x l'ensemble des centres et y leurs excentricités.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n^2)\n",
    "    \"\"\"\n",
    "    excent = {sommet: 0 for sommet in graphe.keys()}\n",
    "    \n",
    "    for sommet in graphe:\n",
    "        excent[sommet] = excentricite_sommet(graphe, sommet)\n",
    "    \n",
    "    return set(sommet for sommet in graphe if excent[sommet] == min(excent.values())), min(excent.values())\n",
    "\n",
    "\n",
    "def diametre(graphe: dict) -> int:\n",
    "    \n",
    "    excent = {sommet: 0 for sommet in graphe.keys()}\n",
    "    \n",
    "    for sommet in graphe:\n",
    "        excent[sommet] = excentricite_sommet(graphe, sommet)\n",
    "    \n",
    "    return set(sommet for sommet in graphe if excent[sommet] == max(excent.values())), max(excent.values())\n",
    "\n",
    "\n",
    "densite = lambda graphe: sum(list(map(len, list(graphe.values()))))/(len(graphe.keys())*(len(graphe.keys()) - 1))\n",
    "\n",
    "\n",
    "def sommets_couvrants(graphe: dict) -> set:\n",
    "    \n",
    "    \"\"\"Chvatal algorithm à la place\"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "        Heuristique de calcul des sommets couvrants (vertex cover).\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - dom (set): l'ensemble (dont on espère minimum) des sommets couvrants l'ensemble des arêtes.\n",
    "            \n",
    "        Complexité:\n",
    "            - O(n^2)\n",
    "    \"\"\"\n",
    "    \n",
    "    deg = {sommet: len(voisins) for sommet, voisins in graphe.items()}\n",
    "    \n",
    "    def procedure(graphe: dict, degres: dict, dom: set) -> set:\n",
    "        \n",
    "        if len(degres) == 0 or max(degres.values()) == 0:\n",
    "            \n",
    "            return dom\n",
    "        \n",
    "        nouveau = max(degres, key=lambda k: degres[k])\n",
    "        \n",
    "        dom = dom.union({nouveau})\n",
    "        \n",
    "        for sommet in degres:\n",
    "            \n",
    "            if nouveau in graphe[sommet]:\n",
    "                \n",
    "                degres[sommet] -= 1\n",
    "                \n",
    "        del degres[nouveau]\n",
    "        \n",
    "        return procedure(graphe, degres, dom)\n",
    "    \n",
    "    return procedure(graphe, deg, set())\n",
    "    \n",
    "\n",
    "def contraction(graphe: dict, s1: int, s2: int) -> dict:\n",
    "    \"\"\"\n",
    "        Algorithme de contraction de deux sommets s1, s2 en un sommet appelé s1/s2.\n",
    "        \n",
    "        Variable:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "            - s1 (int): le sommet n°1 a fusionné.\n",
    "            - s2 (int): le sommet n°2 a fusionné.\n",
    "        \n",
    "        Sorties:\n",
    "            - nouveau_graphe (dict): le graphe contracté.\n",
    "        \n",
    "        Complexité:\n",
    "            - O(?)\n",
    "    \"\"\"\n",
    "    nouveau_graphe = graphe.copy()\n",
    "    \n",
    "    nouveau_graphe[f\"{s1}/{s2}\"] = nouveau_graphe[s1].union(graphe[s2])\n",
    "    \n",
    "    for sommet, voisins in graphe.items():\n",
    "        if s1 in voisins or s2 in voisins:\n",
    "            \n",
    "            nouveau_graphe[sommet] = (nouveau_graphe[sommet].difference({s1, s2})).union({f\"{s1}/{s2}\"})\n",
    "    \n",
    "    del nouveau_graphe[s1]\n",
    "    del nouveau_graphe[s2]\n",
    "    \n",
    "    return nouveau_graphe\n",
    "\n",
    "def contraction_pondere(s1, s2, graphe):\n",
    "    \"\"\"\n",
    "        Algorithme de contraction de deux sommets s1, s2 en un sommet appelé s1/s2 et pour toute arêtes incidentes à \n",
    "        s1 ou s2 on a w(s1/s2, x) = somme(w(s1, x), w(s2, x)).\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {(voisin1, poids1), (voisin2, poids2), ..., (voisinN, poidsN)}}.\n",
    "            - s1 (int): le sommet n°1 a fusionné.\n",
    "            - s2 (int): le sommet n°2 a fusionné.\n",
    "        \n",
    "        Sorties:\n",
    "            - nouveau_graphe (dict): le graphe contracté.\n",
    "        \n",
    "        Complexité:\n",
    "            - O(?)\n",
    "    \"\"\"\n",
    "    sf = f\"{s1}/{s2}\"\n",
    "    nouveau_graphe = graphe.copy()\n",
    "\n",
    "    for sommet in graphe:\n",
    "        nouveau_graphe[sommet] = {arete for arete in graphe[sommet] if arete[0] not in {s1, s2}}\n",
    "    del nouveau_graphe[s1], nouveau_graphe[s2]\n",
    "\n",
    "    for sommet in nouveau_graphe:\n",
    "        nouveau_graphe[sommet] = nouveau_graphe[sommet].union({(sf, sum(list(map(lambda arete: arete[1], {art for art in graphe[sommet] if art[0] in {s1, s2}}))))})\n",
    "\n",
    "    nouveau_graphe[sf] = set()    \n",
    "        \n",
    "    for sommet in nouveau_graphe:\n",
    "        nouveau_graphe[sf] = nouveau_graphe[sf].union({(sommet, arete[1]) for arete in nouveau_graphe[sommet] if arete[0] == sf}) \n",
    "\n",
    "    return nouveau_graphe\n",
    "\n",
    "\n",
    "def stoer_wagner(graphe: dict) -> int:\n",
    "    \"\"\"\n",
    "        Algorithme de Stoer-Wagner afin de trouver la coupe minimale d'un graphe i.e le k pour k-arêtes-connexes.\n",
    "        \n",
    "        Variables:\n",
    "            - graphe (dict): graphe sous la forme de dictionnaire {sommet: {voisin1, voisin2, ..., voisinN}}.\n",
    "        \n",
    "        Sorties:\n",
    "            - (int): La coupe minimale du graphe.\n",
    "        \n",
    "        Complexité:\n",
    "            - O(n^2)\n",
    "    \"\"\"\n",
    "    nouveau_graphe = {sommet: {(voisin, 1) for voisin in voisins} for sommet, voisins in graphe.items()}\n",
    "    \n",
    "    coupe = set()\n",
    "    \n",
    "    while len(nouveau_graphe) > 2:\n",
    "        \n",
    "        graphe_poids = nouveau_graphe.copy()\n",
    "        ensemble = list(sommet for sommet in nouveau_graphe if \"/\" in str(sommet))\n",
    "        ensemble = 1 if ensemble == [] else ensemble[0]\n",
    "        cpt = 0\n",
    "        sommets = []\n",
    "        \n",
    "        while cpt < len(graphe_poids):\n",
    "            \n",
    "            s1, s2, w = None, None, -1\n",
    "                \n",
    "            v1 = ensemble\n",
    "            v2, maximum_local = list(graphe_poids[ensemble])[0]\n",
    "\n",
    "            for arete in graphe_poids[ensemble]:\n",
    "                if arete[1] > maximum_local:\n",
    "                    v2, maximum_local = arete\n",
    "\n",
    "            if maximum_local > w:\n",
    "                s1, s2, w = v1, v2, maximum_local\n",
    "                    \n",
    "            ensemble = f\"{s1}/{s2}\"\n",
    "            sommets.append(s2)\n",
    "            \n",
    "            graphe_poids = contraction_pondere(s1, s2, graphe_poids)\n",
    "            \n",
    "            cpt += 1\n",
    "            \n",
    "        s1, s2 = sommets[-2:]\n",
    "        \n",
    "        s1 = str(s1) if \"/\" in str(s1) else int(s1)\n",
    "        s2 = str(s2) if \"/\" in str(s2) else int(s2)\n",
    "        \n",
    "        nouveau_graphe = contraction_pondere(s1, s2, nouveau_graphe)\n",
    "        coupe = coupe.union({w})\n",
    "    \n",
    "    coupe = coupe.union({max({poids for aretes in nouveau_graphe.values() for _, poids in aretes})})\n",
    "    \n",
    "    return min(coupe)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2939a88b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Transcripts accession</th>\n",
       "      <th>Protein accession</th>\n",
       "      <th>Protein length</th>\n",
       "      <th>Locus tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>seripauperin PAU8</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>851229</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180043.1</td>\n",
       "      <td>NP_009332.1</td>\n",
       "      <td>120.0</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>I</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1466426</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184582.1</td>\n",
       "      <td>NP_878038.1</td>\n",
       "      <td>75.0</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>putative permease SEO1</td>\n",
       "      <td>SEO1</td>\n",
       "      <td>851230</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178208.1</td>\n",
       "      <td>NP_009333.1</td>\n",
       "      <td>593.0</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>11565</td>\n",
       "      <td>11951</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>851232</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179897.1</td>\n",
       "      <td>NP_009335.1</td>\n",
       "      <td>128.0</td>\n",
       "      <td>YAL065C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>12046</td>\n",
       "      <td>12426</td>\n",
       "      <td>I</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>851233</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180042.1</td>\n",
       "      <td>NP_009336.1</td>\n",
       "      <td>126.0</td>\n",
       "      <td>YAL064W-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6443</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>48901</td>\n",
       "      <td>50097</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>mitochondrial 37S ribosomal protein VAR1</td>\n",
       "      <td>VAR1</td>\n",
       "      <td>854586</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009320.1</td>\n",
       "      <td>398.0</td>\n",
       "      <td>Q0140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6445</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>61022</td>\n",
       "      <td>61729</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>intron-encoded endonuclease I-SceI</td>\n",
       "      <td>SCEI</td>\n",
       "      <td>854590</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009324.1</td>\n",
       "      <td>235.0</td>\n",
       "      <td>Q0160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6462</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>73758</td>\n",
       "      <td>74513</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>cytochrome c oxidase subunit 2</td>\n",
       "      <td>COX2</td>\n",
       "      <td>854622</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009326.1</td>\n",
       "      <td>251.0</td>\n",
       "      <td>Q0250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6463</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>74495</td>\n",
       "      <td>75984</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>maturase-like protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854623</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009327.1</td>\n",
       "      <td>472.0</td>\n",
       "      <td>Q0255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6467</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>79213</td>\n",
       "      <td>80022</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>cytochrome c oxidase subunit 3</td>\n",
       "      <td>COX3</td>\n",
       "      <td>854627</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009328.1</td>\n",
       "      <td>269.0</td>\n",
       "      <td>Q0275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6014 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession  Begin    End Chromosome Orientation  \\\n",
       "0     NC_001133.9   1807   2169          I       minus   \n",
       "1     NC_001133.9   2480   2707          I        plus   \n",
       "2     NC_001133.9   7235   9016          I       minus   \n",
       "3     NC_001133.9  11565  11951          I       minus   \n",
       "4     NC_001133.9  12046  12426          I        plus   \n",
       "...           ...    ...    ...        ...         ...   \n",
       "6443  NC_001224.1  48901  50097         MT        plus   \n",
       "6445  NC_001224.1  61022  61729         MT        plus   \n",
       "6462  NC_001224.1  73758  74513         MT        plus   \n",
       "6463  NC_001224.1  74495  75984         MT        plus   \n",
       "6467  NC_001224.1  79213  80022         MT        plus   \n",
       "\n",
       "                                          Name Symbol  Gene ID  \\\n",
       "0                            seripauperin PAU8   PAU8   851229   \n",
       "1                      uncharacterized protein    NaN  1466426   \n",
       "2                       putative permease SEO1   SEO1   851230   \n",
       "3                      uncharacterized protein    NaN   851232   \n",
       "4                      uncharacterized protein    NaN   851233   \n",
       "...                                        ...    ...      ...   \n",
       "6443  mitochondrial 37S ribosomal protein VAR1   VAR1   854586   \n",
       "6445        intron-encoded endonuclease I-SceI   SCEI   854590   \n",
       "6462            cytochrome c oxidase subunit 2   COX2   854622   \n",
       "6463                     maturase-like protein    NaN   854623   \n",
       "6467            cytochrome c oxidase subunit 3   COX3   854627   \n",
       "\n",
       "           Gene Type Transcripts accession Protein accession  Protein length  \\\n",
       "0     protein-coding        NM_001180043.1       NP_009332.1           120.0   \n",
       "1     protein-coding        NM_001184582.1       NP_878038.1            75.0   \n",
       "2     protein-coding        NM_001178208.1       NP_009333.1           593.0   \n",
       "3     protein-coding        NM_001179897.1       NP_009335.1           128.0   \n",
       "4     protein-coding        NM_001180042.1       NP_009336.1           126.0   \n",
       "...              ...                   ...               ...             ...   \n",
       "6443  protein-coding                   NaN       NP_009320.1           398.0   \n",
       "6445  protein-coding                   NaN       NP_009324.1           235.0   \n",
       "6462  protein-coding                   NaN       NP_009326.1           251.0   \n",
       "6463  protein-coding                   NaN       NP_009327.1           472.0   \n",
       "6467  protein-coding                   NaN       NP_009328.1           269.0   \n",
       "\n",
       "      Locus tag  \n",
       "0       YAL068C  \n",
       "1     YAL067W-A  \n",
       "2       YAL067C  \n",
       "3       YAL065C  \n",
       "4     YAL064W-B  \n",
       "...         ...  \n",
       "6443      Q0140  \n",
       "6445      Q0160  \n",
       "6462      Q0250  \n",
       "6463      Q0255  \n",
       "6467      Q0275  \n",
       "\n",
       "[6014 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteome = pd.read_csv(\"Saccharomyces cerevisiae/transcriptome.tsv\", sep=\"\\t\")\n",
    "proteome = proteome[proteome[\"Gene Type\"] == \"protein-coding\"]\n",
    "proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8fd5bc65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(proteome[\"Locus tag\"].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3e6feaf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = {}\n",
    "noms = {}\n",
    "G = {}\n",
    "\n",
    "for i, pid in enumerate(proteome[\"Locus tag\"]):\n",
    "    G[i+1] = set()\n",
    "    indices[pid] = i+1\n",
    "    noms[i+1] = pid\n",
    "\n",
    "with open(\"Saccharomyces cerevisiae/interactome.txt\", \"r\") as fil:\n",
    "    fil.readline()\n",
    "    for ligne in fil:\n",
    "        v1, v2 = ligne[:-1].split(\";\")\n",
    "        if (v1 in indices) and (v2 in indices):\n",
    "            G[indices[v1]].add(indices[v2])\n",
    "            G[indices[v2]].add(indices[v1])\n",
    "        \n",
    "nb_noeuds = len(G.keys())\n",
    "nb_aretes = sum(list(map(len, list(G.values()))))/2\n",
    "\n",
    "#G = {1: {2, 3}, 2: {1, 3}, 3: {1, 2, 4}, 4: {3, 5, 6}, 5: {4, 6}, 6: {5, 4}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3be42141",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'id': 1267611, 'display_name': 'PAU8', 'link': '/locus/S000002142', 'format_name': 'YAL068C'} {'id': 1285991, 'display_name': 'RCK1', 'link': '/locus/S000003126', 'format_name': 'YGL158W'}\n",
      "{'id': 1281159, 'display_name': 'HEK2', 'link': '/locus/S000000128', 'format_name': 'YBL032W'} {'id': 1267611, 'display_name': 'PAU8', 'link': '/locus/S000002142', 'format_name': 'YAL068C'}\n",
      "{'id': 1267611, 'display_name': 'PAU8', 'link': '/locus/S000002142', 'format_name': 'YAL068C'} {'id': 1283876, 'display_name': 'NAM7', 'link': '/locus/S000004685', 'format_name': 'YMR080C'}\n"
     ]
    }
   ],
   "source": [
    "url = lambda pid: f\"https://www.yeastgenome.org/backend/locus/{pid}/interaction_details\"\n",
    "reponse = requests.get(url(\"YAL068C\")).json()\n",
    "for interaction in reponse:\n",
    "    if interaction[\"interaction_type\"] == \"Physical\":\n",
    "        print(interaction[\"locus1\"], interaction[\"locus2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "941161da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YAL068C;YGL158W\n",
      "YAL068C;YBL032W\n",
      "YAL068C;YMR080C\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(\"Saccharomyces cerevisiae/interactome_physique.txt\", \"w\") as filout, open(\"Saccharomyces cerevisiae/non_reference_physique.txt\", \"w\") as fil:\n",
    "    texte, textref = \"\", \"\"\n",
    "    for pid in proteome[\"Locus tag\"]:\n",
    "        reponse = requests.get(url(pid)).json()\n",
    "        try:\n",
    "            for interaction in reponse:\n",
    "                if interaction[\"interaction_type\"] == \"Physical\":\n",
    "                    ss_texte = f\"{interaction['locus1']['format_name']};{interaction['locus2']['format_name']}\\n\" if interaction['locus2']['format_name'] != pid else f\"{interaction['locus2']['format_name']};{interaction['locus1']['format_name']}\\n\"\n",
    "                    texte += ss_texte\n",
    "        except:\n",
    "            textref += f\"{pid}\\n\"\n",
    "    filout.write(texte)\n",
    "    fil.write(textref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2ec5057a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YAL068C;YDL225W\n",
      "YAL068C;YMR044W\n",
      "YAL068C;YDR054C\n",
      "YAL068C;YDR404C\n",
      "YAL068C;YJR065C\n",
      "YAL068C;YMR235C\n",
      "YAL068C;YNL061W\n",
      "YAL068C;YOR074C\n",
      "YAL068C;YOR204W\n",
      "YAL068C;YOR236W\n",
      "YAL068C;YLR142W\n",
      "YAL068C;YBL020W\n",
      "YAL068C;YBL034C\n",
      "YAL068C;YDR531W\n",
      "YAL068C;YFR005C\n",
      "YAL068C;YGR277C\n",
      "YAL068C;YJR045C\n",
      "YAL068C;YDL140C\n",
      "YAL068C;YDR311W\n",
      "YAL068C;YDR510W\n",
      "YAL068C;YER125W\n",
      "YAL068C;YGL073W\n",
      "YAL068C;YIL046W\n",
      "YAL068C;YJL194W\n",
      "YAL068C;YJL203W\n",
      "YAL068C;YKL019W\n",
      "YAL068C;YLR071C\n",
      "YAL068C;YLR293C\n",
      "YAL068C;YLR310C\n",
      "YAL068C;YML130C\n",
      "YAL068C;YOR020C\n",
      "YAL068C;YOR174W\n",
      "YAL068C;YPL082C\n",
      "YAL068C;YPL128C\n",
      "YAL068C;YBR017C\n",
      "YAL068C;YDL225W\n",
      "YAL068C;YDR022C\n",
      "YAL068C;YER110C\n",
      "YAL068C;YHL020C\n",
      "YAL068C;YNCB0010W\n",
      "YAL068C;YJL102W\n",
      "YAL068C;YJR118C\n",
      "YAL068C;YKL117W\n",
      "YAL068C;YKR076W\n",
      "YAL068C;YLR319C\n",
      "YAL068C;YML020W\n",
      "YAL068C;YML101C\n",
      "YAL068C;YMR044W\n",
      "YAL068C;YNL080C\n",
      "YAL068C;YPL029W\n",
      "YAL068C;YPL040C\n",
      "YAL068C;YPR036W\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(\"Saccharomyces cerevisiae/interactome_genetique.txt\", \"w\") as filout, open(\"Saccharomyces cerevisiae/non_reference_genetique.txt\", \"w\") as fil:\n",
    "    texte, textref = \"\", \"\"\n",
    "    for pid in proteome[\"Locus tag\"]:\n",
    "        reponse = requests.get(url(pid)).json()\n",
    "        try:\n",
    "            for interaction in reponse:\n",
    "                if interaction[\"interaction_type\"] == \"Genetic\":\n",
    "                    ss_texte = f\"{interaction['locus1']['format_name']};{interaction['locus2']['format_name']}\\n\" if interaction['locus2']['format_name'] != pid else f\"{interaction['locus2']['format_name']};{interaction['locus1']['format_name']}\\n\"\n",
    "                    texte += ss_texte\n",
    "        except:\n",
    "            textref += f\"{pid}\\n\"\n",
    "    filout.write(texte)\n",
    "    fil.write(textref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9e93da2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55\n"
     ]
    }
   ],
   "source": [
    "with open(\"Saccharomyces cerevisiae/interactome.txt\", \"w\") as filout, open(\"Saccharomyces cerevisiae/non_reference.txt\", \"w\") as fil:\n",
    "    texte, textref = \"\", \"\"\n",
    "    for pid in proteome[\"Locus tag\"]:\n",
    "        reponse = requests.get(url(pid)).json()\n",
    "        try:\n",
    "            ss_texte = f\"{pid};\"\n",
    "            ss_texte += f\"{pid};\".join(list(map(lambda x: x + \"\\n\", [interaction['locus1'][\"format_name\"] if interaction['locus1'][\"format_name\"] != pid else interaction['locus2'][\"format_name\"] for interaction in reponse])))\n",
    "            texte += ss_texte\n",
    "        except:\n",
    "            textref += f\"{pid}\\n\"\n",
    "        break\n",
    "    filout.write(texte)\n",
    "    fil.write(textref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e71dda",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
