{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f20180f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c6f88050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parcours_profondeur(graphe: dict, debut: int) -> list:\n",
    "    \n",
    "    cycle = False\n",
    "    \n",
    "    taille = len(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    pere = [None for _ in range(taille)]\n",
    "    dernier_pred = [None for _ in range(taille)]\n",
    "    \n",
    "    nb_vu = 1\n",
    "    \n",
    "    pile = [debut]\n",
    "    etat[debut - 1] = 0\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        if etat[nb_vu - 1] is not None:\n",
    "            nb_vu += 1\n",
    "        elif pile == []:\n",
    "            pile.append(nb_vu)\n",
    "        while pile != []:\n",
    "            suivant = pile.pop()\n",
    "            if etat[suivant - 1] is None:\n",
    "                etat[suivant - 1] = 0\n",
    "                pere[suivant - 1] = dernier_pred[suivant - 1]\n",
    "                for successeur in graphe[suivant]:\n",
    "                    pile.append(successeur)\n",
    "                    if (etat[successeur - 1] is not None) and (successeur != dernier_pred[suivant - 1]):\n",
    "                        cycle = True\n",
    "                    dernier_pred[successeur - 1] = suivant\n",
    "    return pere\n",
    "\n",
    "def excentricite_sommet(graphe: dict, debut: int) -> list:\n",
    "    \n",
    "    taille = len(graphe.keys())\n",
    "    \n",
    "    etat = [None for _ in range(taille)]\n",
    "    hauteur = [0 for _ in range(taille)]\n",
    "    \n",
    "    file = [debut]\n",
    "    etat[debut - 1] = 0\n",
    "    \n",
    "    nb_vu = 1\n",
    "    \n",
    "    while nb_vu <= taille:\n",
    "        if etat[nb_vu - 1] is not None:\n",
    "            nb_vu += 1\n",
    "        elif file == []:\n",
    "            file.append(nb_vu)\n",
    "            etat[nb_vu - 1] = 0\n",
    "        while file != []:\n",
    "            suivant = file[0]\n",
    "            file = file[1:] if len(file) > 1 else []\n",
    "            for successeur in graphe[suivant]:\n",
    "                if etat[successeur - 1] is None:\n",
    "                    file.append(successeur)\n",
    "                    etat[successeur - 1] = 0\n",
    "                    hauteur[successeur - 1] = hauteur[suivant - 1] + 1\n",
    "                    \n",
    "    return max(hauteur)\n",
    "\n",
    "\"\"\"Articulation => Médiateurs, Centres => ?, Degre, Diamètre, Graphe complémentaire et stable => Protéine localisée !=\n",
    "Traité les composantes connexes séparemment, Isthme => ?\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2939a88b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Transcripts accession</th>\n",
       "      <th>Protein accession</th>\n",
       "      <th>Protein length</th>\n",
       "      <th>Locus tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>seripauperin PAU8</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>851229</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180043.1</td>\n",
       "      <td>NP_009332.1</td>\n",
       "      <td>120.0</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>I</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1466426</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184582.1</td>\n",
       "      <td>NP_878038.1</td>\n",
       "      <td>75.0</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>putative permease SEO1</td>\n",
       "      <td>SEO1</td>\n",
       "      <td>851230</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178208.1</td>\n",
       "      <td>NP_009333.1</td>\n",
       "      <td>593.0</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>11565</td>\n",
       "      <td>11951</td>\n",
       "      <td>I</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>851232</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179897.1</td>\n",
       "      <td>NP_009335.1</td>\n",
       "      <td>128.0</td>\n",
       "      <td>YAL065C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>12046</td>\n",
       "      <td>12426</td>\n",
       "      <td>I</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>851233</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180042.1</td>\n",
       "      <td>NP_009336.1</td>\n",
       "      <td>126.0</td>\n",
       "      <td>YAL064W-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6443</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>48901</td>\n",
       "      <td>50097</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>mitochondrial 37S ribosomal protein VAR1</td>\n",
       "      <td>VAR1</td>\n",
       "      <td>854586</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009320.1</td>\n",
       "      <td>398.0</td>\n",
       "      <td>Q0140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6445</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>61022</td>\n",
       "      <td>61729</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>intron-encoded endonuclease I-SceI</td>\n",
       "      <td>SCEI</td>\n",
       "      <td>854590</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009324.1</td>\n",
       "      <td>235.0</td>\n",
       "      <td>Q0160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6462</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>73758</td>\n",
       "      <td>74513</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>cytochrome c oxidase subunit 2</td>\n",
       "      <td>COX2</td>\n",
       "      <td>854622</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009326.1</td>\n",
       "      <td>251.0</td>\n",
       "      <td>Q0250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6463</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>74495</td>\n",
       "      <td>75984</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>maturase-like protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854623</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009327.1</td>\n",
       "      <td>472.0</td>\n",
       "      <td>Q0255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6467</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>79213</td>\n",
       "      <td>80022</td>\n",
       "      <td>MT</td>\n",
       "      <td>plus</td>\n",
       "      <td>cytochrome c oxidase subunit 3</td>\n",
       "      <td>COX3</td>\n",
       "      <td>854627</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_009328.1</td>\n",
       "      <td>269.0</td>\n",
       "      <td>Q0275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6014 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession  Begin    End Chromosome Orientation  \\\n",
       "0     NC_001133.9   1807   2169          I       minus   \n",
       "1     NC_001133.9   2480   2707          I        plus   \n",
       "2     NC_001133.9   7235   9016          I       minus   \n",
       "3     NC_001133.9  11565  11951          I       minus   \n",
       "4     NC_001133.9  12046  12426          I        plus   \n",
       "...           ...    ...    ...        ...         ...   \n",
       "6443  NC_001224.1  48901  50097         MT        plus   \n",
       "6445  NC_001224.1  61022  61729         MT        plus   \n",
       "6462  NC_001224.1  73758  74513         MT        plus   \n",
       "6463  NC_001224.1  74495  75984         MT        plus   \n",
       "6467  NC_001224.1  79213  80022         MT        plus   \n",
       "\n",
       "                                          Name Symbol  Gene ID  \\\n",
       "0                            seripauperin PAU8   PAU8   851229   \n",
       "1                      uncharacterized protein    NaN  1466426   \n",
       "2                       putative permease SEO1   SEO1   851230   \n",
       "3                      uncharacterized protein    NaN   851232   \n",
       "4                      uncharacterized protein    NaN   851233   \n",
       "...                                        ...    ...      ...   \n",
       "6443  mitochondrial 37S ribosomal protein VAR1   VAR1   854586   \n",
       "6445        intron-encoded endonuclease I-SceI   SCEI   854590   \n",
       "6462            cytochrome c oxidase subunit 2   COX2   854622   \n",
       "6463                     maturase-like protein    NaN   854623   \n",
       "6467            cytochrome c oxidase subunit 3   COX3   854627   \n",
       "\n",
       "           Gene Type Transcripts accession Protein accession  Protein length  \\\n",
       "0     protein-coding        NM_001180043.1       NP_009332.1           120.0   \n",
       "1     protein-coding        NM_001184582.1       NP_878038.1            75.0   \n",
       "2     protein-coding        NM_001178208.1       NP_009333.1           593.0   \n",
       "3     protein-coding        NM_001179897.1       NP_009335.1           128.0   \n",
       "4     protein-coding        NM_001180042.1       NP_009336.1           126.0   \n",
       "...              ...                   ...               ...             ...   \n",
       "6443  protein-coding                   NaN       NP_009320.1           398.0   \n",
       "6445  protein-coding                   NaN       NP_009324.1           235.0   \n",
       "6462  protein-coding                   NaN       NP_009326.1           251.0   \n",
       "6463  protein-coding                   NaN       NP_009327.1           472.0   \n",
       "6467  protein-coding                   NaN       NP_009328.1           269.0   \n",
       "\n",
       "      Locus tag  \n",
       "0       YAL068C  \n",
       "1     YAL067W-A  \n",
       "2       YAL067C  \n",
       "3       YAL065C  \n",
       "4     YAL064W-B  \n",
       "...         ...  \n",
       "6443      Q0140  \n",
       "6445      Q0160  \n",
       "6462      Q0250  \n",
       "6463      Q0255  \n",
       "6467      Q0275  \n",
       "\n",
       "[6014 rows x 13 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteome = pd.read_csv(\"Saccharomyces cerevisiae/transcriptome.tsv\", sep=\"\\t\")\n",
    "proteome = proteome[proteome[\"Gene Type\"] == \"protein-coding\"]\n",
    "proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8fd5bc65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(proteome[\"Locus tag\"].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7c0be1ab",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YNR020C 78\n"
     ]
    }
   ],
   "source": [
    "url = lambda tag, k: f\"http://webservice.thebiogrid.org/interactions?geneList={tag}&format=json&taxId=559292&accesskey={k}\"\n",
    "cle = \"47c1bde1e3d0203e63756e71e9ca4f81\"\n",
    "reponse = requests.get(url(\"Q0275\", cle)).json()\n",
    "print(reponse[\"2513187\"][\"SYSTEMATIC_NAME_A\"], len(reponse.keys()))\n",
    "\n",
    "# vérifier avec biogrid en passant l'id 854627"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "edb99b64",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# +40 min de lancement faites attention\n",
    "with open(\"Saccharomyces cerevisiae/interactome.txt\", \"w\") as filout, open(\"Saccharomyces cerevisiae/non_reference.txt\", \"w\") as fil:\n",
    "    texte = \"PID(1);PID(2)\\n\"\n",
    "    textref = \"\"\n",
    "    for pid in proteome[\"Locus tag\"]:\n",
    "        reponse = requests.get(url(pid, cle)).json()\n",
    "        try:\n",
    "            ss_texte = f\"{pid};\"\n",
    "            ss_texte += f\"{pid};\".join(list(map(lambda x: x + \"\\n\", [reponse[interaction][\"SYSTEMATIC_NAME_A\"] if reponse[interaction][\"SYSTEMATIC_NAME_A\"] != pid else reponse[interaction][\"SYSTEMATIC_NAME_B\"] for interaction in reponse.keys()]))\n",
    "    )\n",
    "            texte += ss_texte\n",
    "        except:\n",
    "            textref += f\"{pid}\\n\"\n",
    "    filout.write(texte)\n",
    "    fil.write(textref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "955b9794",
   "metadata": {},
   "outputs": [],
   "source": [
    "#18h10 -> 19h18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3e6feaf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = {}\n",
    "noms = {}\n",
    "G = {}\n",
    "\n",
    "for i, pid in enumerate(proteome[\"Locus tag\"]):\n",
    "    G[i+1] = set()\n",
    "    indices[pid] = i+1\n",
    "    noms[i+1] = pid\n",
    "\n",
    "with open(\"Saccharomyces cerevisiae/interactome.txt\", \"r\") as fil:\n",
    "    fil.readline()\n",
    "    for ligne in fil:\n",
    "        v1, v2 = ligne[:-1].split(\";\")\n",
    "        if (v1 in indices) and (v2 in indices):\n",
    "            G[indices[v1]].add(indices[v2])\n",
    "            G[indices[v2]].add(indices[v1])\n",
    "         \n",
    "        \n",
    "nb_noeuds = len(G.keys())\n",
    "nb_aretes = sum(list(map(len, list(G.values()))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b762cf06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6014 1216256\n"
     ]
    }
   ],
   "source": [
    "print(nb_noeuds, nb_aretes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed3d7aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
